server/bearer_auth.go

```diff
@@ -0,0 +1,110 @@
+package server
+
+import (
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/dgrijalva/jwt-go"
+	natsjwt "github.com/nats-io/jwt"
+)
+
+type BearerAuth struct {
+	server    *Server
+	publicKey *rsa.PublicKey
+}
+
+func BearerAuthFactory(s *Server) (*BearerAuth, error) {
+	auth := &BearerAuth{
+		server: s,
+	}
+	err := auth.readPublicKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read JWT_SIGNER_PUBLIC_KEY from environment")
+	}
+	return auth, nil
+}
+
+func (bearer *BearerAuth) readPublicKey() error {
+	jwtPublicKeyPEM := strings.Replace(os.Getenv("JWT_SIGNER_PUBLIC_KEY"), `\n`, "\n", -1)
+	publicKey, err := jwt.ParseRSAPublicKeyFromPEM([]byte(jwtPublicKeyPEM))
+	if err != nil {
+		return err
+	}
+	bearer.publicKey = publicKey
+	return nil
+}
+
+func (bearer *BearerAuth) Check(c ClientAuthentication) bool {
+	bearerToken := c.GetOpts().JWT
+	jwtToken, err := jwt.Parse(bearerToken, func(_jwtToken *jwt.Token) (interface{}, error) {
+		if _, ok := _jwtToken.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("failed to parse bearer authorization; unexpected signing alg: %s", _jwtToken.Method.Alg())
+		}
+		return bearer.publicKey, nil
+	})
+	if err != nil {
+		bearer.server.Debugf(fmt.Sprintf("failed to parse bearer authorization; %s", err.Error()))
+		return false
+	}
+
+	bearer.server.Debugf(fmt.Sprintf("parsed bearer authorization: %s\n; client authentication: %s", jwtToken.Claims, c))
+	claims, claimsOk := jwtToken.Claims.(jwt.MapClaims)
+	if !claimsOk {
+		bearer.server.Warnf(fmt.Sprintf("no claims present in verified JWT; %s", err.Error()))
+		return false
+	}
+
+	permissions := &Permissions{}
+	if permissionsClaim, permissionsClaimOk := claims["permissions"].(map[string]interface{}); permissionsClaimOk {
```

Comment:
FIXME... the `permissions` claim should be tucked underneath the `nats` claim.

i.e., here is a current JWT payload as generated by `nsc`:
```
{
  "jti": "7WA3QTYWKVEEETMNBJENR34L5SQQYYCE7QB5H7MLEIDBYHX5XGUA",
  "iat": 1576587866,
  "iss": "OCZM6CMZMDLBEECLABZZOLKF7RFTUOWWEL3BHBOKL6GS6VMM56V6X3BB",
  "name": "A",
  "sub": "ABPAHQBU2XAHCENGXGV7NITB4PR2CNEMO3RZKTYJXA3WUR3Q56Y7QUEU",
  "type": "account",
  "nats": {
    "limits": {
      "subs": -1,
      "conn": -1,
      "leaf": -1,
      "imports": -1,
      "exports": -1,
      "data": -1,
      "payload": -1,
      "wildcards": true
    }
  }
}
```

I propose that under this bearer permissions strategy, the `nats` claim encoded in vended JWTs should look like this (other standard JWT claims omitted for brevity):
```
  "nats": {
    "limits": {
      "subs": -1,
      "conn": -1,
      "leaf": -1,
      "imports": -1,
      "exports": -1,
      "data": -1,
      "payload": -1,
      "wildcards": true
    },
    "permissions": {
      "publish": {
        "allow": [],
        "deny": []
      },
      "subscribe": {
        "allow": [],
        "deny": []
      },
      "responses": {
        "max": 1,
        "ttl": 120000000000
      },
    }
  }
}
```